### PR TITLE
Capture telemetry in headers for FLW and Multiple WPJ

### PR DESCRIPTION
**Pull request by shahzaibj**
Related: https://github.com/AzureAD/ad-accounts-for-android/pull/1912

### Description
I would suggest reading the description in the above linked related PR to get a complete sense of what's going on.

In a nutshell, we've made some changes here to facilitate us to capture header based telemetry around FLW and multiple Workplace Join. This is based on Olga's design PR here: https://dev.azure.com/IdentityDivision/DevEx/_git/AuthLibrariesApiReview/pullrequest/5168

Basically, we've separated Android platform fields with shared platform fields we're going to have with Android and iOS.

There are going to be 3 sets of platform fields:

- Android only platform fields
- Android and iOS shared platform fields for FLW
- Android and iOS shared platform fields for Multiple WPJ

The shared platform fields will always come first in the schema and Android only will come second. 

In addition to that, this PR also contains a bunch of utilities class around http request interception, examination, and re-writing to facilitate tests that involve us to do something with the network request. For example, following interceptors are introduced:

- HttpRequestExaminer - allows to intercept and examine the request URL and headers of an HTTP request
- HttpRequestRewriter - allows to intercept and re-write headers or URL or body of an HTTP request
- HttpResponseRewriter - allows to intercept and re-write the response of an HTTP request

### CODE REVIEW

File: common4j/build.gradle [hunk 1]
```diff
@@ -187,10 +213,15 @@ dependencies {
...
+
+    testFixturesCompileOnly 'com.github.spotbugs:spotbugs-annotations:3.1.12'
+    testFixturesCompileOnly "org.projectlombok:lombok:$rootProject.ext.lombokVersion"
+    testFixturesAnnotationProcessor "org.projectlombok:lombok:$rootProject.ext.lombokVersion"
+    resolvableTestFixturesImplementation 'org.robolectric:junit:4.5.1'
 }
 
-sourceCompatibility = "1.7"
-targetCompatibility = "1.7"
+sourceCompatibility = "1.8"
```
Comment (iamgusain) [inline common4j/build.gradle:223]: Why do we need this change?

File: common4j/src/main/com/microsoft/identity/common/java/interfaces/IPerSeparatorNameValueStore.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import com.microsoft.identity.common.java.util.ported.Predicate;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+import edu.umd.cs.findbugs.annotations.Nullable;
+import lombok.NonNull;
+
+public interface IPerSeparatorNameValueStore<S, T> {
```
Comment (iamgusain) [inline common4j/src/main/com/microsoft/identity/common/java/interfaces/IPerSeparatorNameValueStore.java:34]: please add javadoc to define what this interface is for

File: common4j/src/main/com/microsoft/identity/common/java/eststelemetry/SchemaConstants.java [hunk 3]
```diff
@@ -54,6 +56,17 @@ public static final class Key {
         public static final String FRT_STATUS = TelemetryEventStrings.Key.FRT_STATUS;
         public static final String MRRT_STATUS = TelemetryEventStrings.Key.MRRT_STATUS;
         public static final String ALL_TELEMETRY_DATA_SENT = "is_all_telemetry_data_sent";
+
+        // flw and multiple reg fields
+        public static final String IS_SHARED_DEVICE = "isSharedScenario";
```
Comment (iamgusain) [inline common4j/src/main/com/microsoft/identity/common/java/eststelemetry/SchemaConstants.java:61]: "isSharedDevice"

File: common4j/src/main/com/microsoft/identity/common/java/eststelemetry/SchemaConstants.java [hunk 4]
```diff
@@ -54,6 +56,17 @@ public static final class Key {
...
+        public static final String REG_TYPE = "reg_type";
```
Comment (iamgusain) [inline common4j/src/main/com/microsoft/identity/common/java/eststelemetry/SchemaConstants.java:62]: can we add comments to tell what each of these fields is designed to capture.

**Review by iamgusain — APPROVED** [review]

**Comment by shahzaibj** [comment]
This got merged as part of other small broken down PRs.

**Pull request closed without merging**